extensions/src/org/pentaho/platform/plugin/services/security/userrole/jdbc/JdbcUserRoleListService.java

```diff
@@ -146,6 +152,14 @@ public String getAllUsernamesQuery() {
       }
 
     }
+
+    // Now add extra role if it does not exist in the list
+    for ( String extraRole : extraRoles ) {
+      if ( !roles.contains( extraRole ) ) {
```

Comment:
Strongly suggest a LinkedHashSet instead of this. For LDAP, the list of roles will be potentially very long. This is an o^n search across all the roles for the extra roles.

Something like:
     LinkedHashSet<String> roles = new LinkedHashSet<String>( allAuths.size() );
     for ( GrantedAuthority role : allAuths ) {
       if ( roleMapper != null ) {
         roles.add( roleMapper.toPentahoRole( role.getAuthority() ) );
       } else {
         roles.add( role.getAuthority() );
       }
      }

     // Now add extra role if it does not exist in the list
     for ( String extraRole : extraRoles ) {
       roles.add( extraRoles );
     }
     return new ArrayList<String>(roles);

This is just my $0.02 though...